amd/src/courseformat/tabtree.js

```diff
@@ -0,0 +1,150 @@
+        for (let i = 0; i < tabs.length - 1; i++) { // Don't count last "add section" tab.
+            let tab = tabs.item(i);
+            let id = tab.querySelector("a div").dataset.itemid;
+            let classes = tab.querySelector("a").classList.value;
+            this.tabs[id] = tab;
+            if (classes.indexOf(this.classes.ACTIVETAB) !== -1) {
+                this.activetab = i;
+            }
+        }
+
+        const childtabs = this.getElements(this.selectors.CHILDTAB);
+        for (let i = 0; i < childtabs.length - 1; i++) { // Don't count last "add section" tab.
+            let tab = childtabs.item(i);
+            let id = tab.querySelector("a div").dataset.itemid;
+            this.childtabs[id] = tab;
+        }
+    }
+
+    /**
+     * Refresh the section tabs.
+     *
+     * @param {object} param
+     * @param {Object} param.element
+     */
+    _refreshCourseSectionTabs({element}) {
+        // Do things that make the first row tabs match firstsectionlist.
+        const toptabslist = element.firstsectionlist ?? [];
+        let toptabs = this.element.querySelector('ul:first-of-type');
+        this._fixOrder(toptabs, toptabslist, this.tabs);
+
+        // And the second row tabs match secondsectionlist.
+        const childtabslist = element.secondsectionlist ?? [];
+        let childtabs = this.element.querySelector('ul:nth-of-type(2)');
+        this._fixOrder(childtabs, childtabslist[this.activetab], this.childtabs);
+    }
+
+    /**
+     * Fix/reorder the section or cms order.
+     *
+     * @param {Element} container the HTML element to reorder.
+     * @param {Array} neworder an array with the ids order
+     * @param {Array} allitems the list of html elements that can be placed in the container
+     */
+    _fixOrder(container, neworder, allitems) {
+
+        // Empty lists should not be visible.
+        if (!neworder.length) {
+            container.classList.add('hidden');
+            container.innerHTML = '';
+            return;
+        }
+
+        // Grant the list is visible (in case it was empty).
+        container.classList.remove('hidden');
+
+        // Move the elements in order at the beginning of the list.
+        neworder.forEach((itemid, index) => {
+            const item = allitems[itemid];
+            // Get the current element at that position.
+            const currentitem = container.children[index];
+            if (currentitem === undefined) {
+                container.append(item);
+                return;
+            }
+            if (currentitem !== item) {
+                container.insertBefore(item, currentitem);
+            }
+        });
+        // Remove the remaining elements.
+        // Probably not necessary as we are not removing anything. And we don't want the "Add" blown away.
+        /*
+        while (container.children.length > neworder.length) {
+            container.removeChild(container.lastChild);
+        }
+         */
```

Comment:
This looks like it should remove unneeded tabs.
```suggestion
        while (container.children.length - 1 > neworder.length) {
            container.removeChild(container.lastElementChild.previousElementSibling);
```